File: data-pipeline-libraries-java/data-pipeline-core-java/src/main/java/com/enrichmeai/culvert/governance/PiiMaskingGovernancePolicy.java
package com.enrichmeai.culvert.governance;

import com.enrichmeai.culvert.contracts.GovernancePolicy;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * A structural {@link GovernancePolicy} that identifies PII fields by
 * column-name membership and/or regex pattern matching, then applies a
 * configured {@link MaskingPolicy}.
 *
 * <h2>Matching rules (applied in order; first match wins)</h2>
 * <ol>
 *   <li><strong>Column-name set</strong> — exact, case-sensitive comparison
 *       against the {@code piiColumns} set supplied at construction. E.g.
 *       {@code {"email", "ssn", "phone"}}.</li>
 *   <li><strong>Regex patterns</strong> — each pattern in {@code piiPatterns}
 *       is tested with {@link java.util.regex.Matcher#matches()} (full-field-name
 *       match). E.g. {@code [".*_pii$", ".*_secret$"]}.</li>
 * </ol>
 *
 * <p>A field that matches either rule is classified as
 * {@link DataClassification#RESTRICTED}; all other fields are
 * {@link DataClassification#INTERNAL}.
 *
 * <h2>Masking policy resolution</h2>
 * <p>When a field matches:
 * <ol>
 *   <li>If the optional {@code columnOverrides} map contains an entry for that
 *       field name, that {@link MaskingPolicy} is used.</li>
 *   <li>Otherwise the {@code defaultMaskingPolicy} is used.</li>
 * </ol>
 * The {@code columnOverrides} map only overrides the <em>policy</em> for
 * already-matched fields; a key present only in the override map (not in the
 * column set and not matching any pattern) does NOT by itself make a field
 * match. Column matching and policy selection are independent.
 *
 * <h2>Scope cap</h2>
 * <p>This implementation is <strong>structural only</strong>: it inspects
 * field names, not field values. Tag-based policy resolution (Dataplex tags,
 * Cloud DLP per-cell inspection, Purview sensitivity labels) is explicitly
 * <strong>out of scope</strong>. Those paths require a cloud SDK and belong in
 * a separate implementation. Do not add them here.
 *
 * <h2>Zero cloud / Beam imports</h2>
 * <p>This class imports only {@code java.*} and {@code com.enrichmeai.culvert.*}.
 * CI asserts this invariant via an import-line grep (see
 * {@code PiiMaskingGovernancePolicyTest}).
 *
 * <h2>Usage</h2>
 * <pre>{@code
 * PiiMaskingGovernancePolicy policy = PiiMaskingGovernancePolicy.builder()
 *         .piiColumns(Set.of("email", "ssn", "phone"))
 *         .piiPatterns(List.of(".*_pii$", ".*_secret$"))
 *         .defaultMaskingPolicy(new MaskingPolicy(MaskingStrategy.FULL, "***", ""))
 *         .columnOverride("phone", new MaskingPolicy(MaskingStrategy.PARTIAL, "*", ""))
 *         .build();
 *
 * // Or pass directly to DataQualityTransform:
 * DataQualityTransform<Map<String, Object>> dq =
 *         new DataQualityTransform<>(schema, Function.identity(), policy);
 * }</pre>
 *
 * @since Sprint 14 / T14.4 / issue #76
 */
public final class PiiMaskingGovernancePolicy implements GovernancePolicy {

    private final Set<String>       piiColumns;
    private final List<Pattern>     compiledPatterns;
    private final MaskingPolicy     defaultMaskingPolicy;
    private final Map<String, MaskingPolicy> columnOverrides;

    // -----------------------------------------------------------------------
    // Constructor (use builder for readability)
    // -----------------------------------------------------------------------

    /**
     * Creates a new policy.
     *
     * @param piiColumns           Exact column names that are PII.
     *                             Case-sensitive. May be empty.
     * @param piiPatterns          Regex patterns tested against field names
     *                             (full match via {@code matches()}). May be
     *                             empty. Compiled once at construction time.
     * @param defaultMaskingPolicy The masking policy applied when a field
     *                             matches and no per-column override exists.
     *                             Must not be null.
     * @param columnOverrides      Optional per-column policy overrides.
     *                             Only applied to <em>matched</em> fields.
     *                             May be null or empty.
     * @throws NullPointerException     if any of the required args is null.
     * @throws java.util.regex.PatternSyntaxException if any pattern is invalid.
     */
    public PiiMaskingGovernancePolicy(
            Set<String>              piiColumns,
            List<String>             piiPatterns,
            MaskingPolicy            defaultMaskingPolicy,
            Map<String, MaskingPolicy> columnOverrides) {

        Objects.requireNonNull(piiColumns,           "piiColumns must not be null");
        Objects.requireNonNull(piiPatterns,          "piiPatterns must not be null");
        Objects.requireNonNull(defaultMaskingPolicy, "defaultMaskingPolicy must not be null");

        this.piiColumns           = Set.copyOf(piiColumns);
        this.compiledPatterns     = piiPatterns.stream()
                                        .map(Pattern::compile)
                                        .collect(Collectors.toUnmodifiableList());
        this.defaultMaskingPolicy = defaultMaskingPolicy;
        this.columnOverrides      = (columnOverrides != null)
                ? Map.copyOf(columnOverrides)
                : Collections.emptyMap();
    }

    // -----------------------------------------------------------------------
    // GovernancePolicy implementation
    // -----------------------------------------------------------------------

    /**
     * Returns {@link DataClassification#RESTRICTED} for any field whose name
     * is in the PII column set or matches one of the regex patterns;
     * {@link DataClassification#INTERNAL} otherwise.
     *
     * @param field The field name. Must not be null.
     * @param table The table/entity name. Accepted but not used for
     *              structural matching (reserved for future tag-based
     *              delegation).
     * @return {@code RESTRICTED} or {@code INTERNAL}.
     */
    @Override
    public DataClassification classify(String field, String table) {
        return isPii(field) ? DataClassification.RESTRICTED : DataClassification.INTERNAL;
    }

    /**
     * Returns the masking policy for {@code field}, or {@link Optional#empty()}
     * if the field is not PII.
     *
     * <p>When a field matches, the per-column override (if present) takes
     * precedence over the default masking policy.
     *
     * @param field The field name. Must not be null.
     * @param table The table/entity name (not used for structural matching).
     * @return The resolved {@link MaskingPolicy}, or empty.
     */
    @Override
    public Optional<MaskingPolicy> maskingFor(String field, String table) {
        if (!isPii(field)) return Optional.empty();
        MaskingPolicy override = columnOverrides.get(field);
        return Optional.of(override != null ? override : defaultMaskingPolicy);
    }

    /**
     * Returns {@link Optional#empty()} — this policy does not manage retention.
     */
    @Override
    public Optional<com.enrichmeai.culvert.governance.RetentionPolicy> retentionFor(String table) {
        return Optional.empty();
    }

    // -----------------------------------------------------------------------
    // Accessors (for tests and composition)
    // -----------------------------------------------------------------------

    /** Returns an unmodifiable view of the explicit PII column-name set. */
    public Set<String> getPiiColumns() {
        return piiColumns;
    }

    /** Returns the default masking policy. */
    public MaskingPolicy getDefaultMaskingPolicy() {
        return defaultMaskingPolicy;
    }

    /** Returns an unmodifiable view of the per-column override map. */
    public Map<String, MaskingPolicy> getColumnOverrides() {
        return columnOverrides;
    }

    // -----------------------------------------------------------------------
    // Internal: shared match predicate
    // -----------------------------------------------------------------------

    /**
     * Returns {@code true} iff {@code fieldName} is in the explicit column set
     * <em>or</em> matches at least one compiled regex pattern.
     *
     * <p>This is the single source of truth for both {@link #classify} and
     * {@link #maskingFor} — keeping both methods consistent by construction.
     */
    private boolean isPii(String fieldName) {
        if (piiColumns.contains(fieldName)) return true;
        for (Pattern p : compiledPatterns) {
            if (p.matcher(fieldName).matches()) return true;
        }
        return false;
    }

    // -----------------------------------------------------------------------
    // Builder
    // -----------------------------------------------------------------------

    /** Returns a new builder for {@link PiiMaskingGovernancePolicy}. */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Fluent builder for {@link PiiMaskingGovernancePolicy}.
     *
     * <p>All fields are optional except {@code defaultMaskingPolicy}, which
     * must be set before calling {@link #build()}.
     */
    public static final class Builder {

        private Set<String>                piiColumns           = Collections.emptySet();
        private List<String>               piiPatterns          = Collections.emptyList();
        private MaskingPolicy              defaultMaskingPolicy = null;
        private final Map<String, MaskingPolicy> columnOverrides =
                new java.util.LinkedHashMap<>();

        private Builder() {}

        /** Sets the explicit PII column-name set (exact, case-sensitive). */
        public Builder piiColumns(Set<String> columns) {
            this.piiColumns = Objects.requireNonNull(columns, "columns must not be null");
            return this;
        }

        /** Sets the regex pattern list matched against field names. */
        public Builder piiPatterns(List<String> patterns) {
            this.piiPatterns = Objects.requireNonNull(patterns, "patterns must not be null");
            return this;
        }

        /** Sets the default masking policy (required). */
        public Builder defaultMaskingPolicy(MaskingPolicy policy) {
            this.defaultMaskingPolicy = Objects.requireNonNull(policy, "policy must not be null");
            return this;
        }

        /** Adds a per-column override entry. */
        public Builder columnOverride(String column, MaskingPolicy policy) {
            columnOverrides.put(
                    Objects.requireNonNull(column, "column must not be null"),
                    Objects.requireNonNull(policy, "policy must not be null"));
            return this;
        }

        /** Builds the policy. Throws if {@code defaultMaskingPolicy} was not set. */
        public PiiMaskingGovernancePolicy build() {
            if (defaultMaskingPolicy == null) {
                throw new IllegalStateException("defaultMaskingPolicy must be set");
            }
            return new PiiMaskingGovernancePolicy(
                    piiColumns, piiPatterns, defaultMaskingPolicy,
                    columnOverrides.isEmpty() ? null : Map.copyOf(columnOverrides));
        }
    }
}

File: data-pipeline-libraries/data-pipeline-core/src/data_pipeline_core/governance_api/pii_masking_governance_policy.py
"""PiiMaskingGovernancePolicy — structural GovernancePolicy implementation.

Identifies PII fields by column-name membership and/or regex pattern
matching, then applies a configured MaskingPolicy.

Mirrors Java ``com.enrichmeai.culvert.governance.PiiMaskingGovernancePolicy``
(Sprint 14 / T14.4 / issue #76).

Matching rules (applied in order; first match wins)
----------------------------------------------------
1. **Column-name set** — exact, case-sensitive comparison against the
   ``pii_columns`` set supplied at construction.
   E.g. ``{"email", "ssn", "phone"}``.
2. **Regex patterns** — each pattern in ``pii_patterns`` is tested with
   ``re.fullmatch()`` (anchored full-field-name match, mirroring Java
   ``Matcher.matches()``).  E.g. ``[".*_pii$", ".*_secret$"]``.

Classification
--------------
A field that matches either rule → ``DataClassification.RESTRICTED``.
All other fields → ``DataClassification.INTERNAL``.

Masking policy resolution
--------------------------
When a field matches:
1. If ``column_overrides`` contains an entry for that field name, that
   MaskingPolicy is used.
2. Otherwise ``default_masking_policy`` is used.

``column_overrides`` only overrides the *policy* for already-matched
fields. A key present only in the override map (not in the column set
and not matching any pattern) does NOT by itself make a field match.

Scope cap
---------
Structural only — inspects field names, not values. Tag-based policy
resolution (Dataplex, Cloud DLP, Purview) is out of scope. Do not add
cloud SDK imports here.
"""

from __future__ import annotations

import re
from typing import Dict, FrozenSet, List, Mapping, Optional, Sequence

from data_pipeline_core.governance_api.classification import DataClassification
from data_pipeline_core.governance_api.policies import MaskingPolicy, RetentionPolicy


class PiiMaskingGovernancePolicy:
    """A structural :class:`GovernancePolicy` for PII masking.

    Parameters
    ----------
    pii_columns:
        Exact column names that are PII (case-sensitive).  May be empty.
    pii_patterns:
        Regex pattern strings tested against field names via
        ``re.fullmatch()`` (anchored, mirrors Java ``Matcher.matches()``).
        May be empty.  Patterns are compiled once at construction time.
    default_masking_policy:
        Applied when a field matches and no per-column override exists.
        Required.
    column_overrides:
        Optional per-column policy overrides, applied only to *matched*
        fields.

    Raises
    ------
    ValueError
        If ``default_masking_policy`` is ``None``.
    re.error
        If any pattern in ``pii_patterns`` is invalid.
    """

    # Mirrors Java PiiMaskingGovernancePolicy (PiiMaskingGovernancePolicy.java:73)

    def __init__(
        self,
        pii_columns: Optional[FrozenSet[str] | frozenset | set] = None,
        pii_patterns: Optional[Sequence[str]] = None,
        default_masking_policy: Optional[MaskingPolicy] = None,
        column_overrides: Optional[Mapping[str, MaskingPolicy]] = None,
    ) -> None:
        if default_masking_policy is None:
            raise ValueError("default_masking_policy must not be None")

        self._pii_columns: FrozenSet[str] = frozenset(pii_columns) if pii_columns else frozenset()
        self._compiled_patterns: List[re.Pattern[str]] = [
            re.compile(p) for p in (pii_patterns or [])
        ]
        self._default_masking_policy: MaskingPolicy = default_masking_policy
        self._column_overrides: Dict[str, MaskingPolicy] = dict(column_overrides) if column_overrides else {}

    # ------------------------------------------------------------------
    # GovernancePolicy implementation
    # ------------------------------------------------------------------

    def classify(self, field: str, table: str) -> DataClassification:
        """Return RESTRICTED if *field* is PII, INTERNAL otherwise.

        Mirrors Java ``PiiMaskingGovernancePolicy.classify``
        (PiiMaskingGovernancePolicy.java:137-139).
        """
        return DataClassification.RESTRICTED if self._is_pii(field) else DataClassification.INTERNAL

    def masking_for(self, field: str, table: str) -> Optional[MaskingPolicy]:
        """Return the masking policy for *field*, or None if not PII.

        Column override (if present) takes precedence over the default.

        Mirrors Java ``PiiMaskingGovernancePolicy.maskingFor``
        (PiiMaskingGovernancePolicy.java:152-157).
        """
        if not self._is_pii(field):
            return None
        return self._column_overrides.get(field, self._default_masking_policy)

    def retention_for(self, table: str) -> Optional[RetentionPolicy]:
        """Always returns None — this policy does not manage retention.

        Mirrors Java ``PiiMaskingGovernancePolicy.retentionFor``
        (PiiMaskingGovernancePolicy.java:163-165).
        """
        return None

    # ------------------------------------------------------------------
    # Accessors (for tests and composition)
    # ------------------------------------------------------------------

    @property
    def pii_columns(self) -> FrozenSet[str]:
        """Unmodifiable view of the explicit PII column-name set."""
        return self._pii_columns

    @property
    def default_masking_policy(self) -> MaskingPolicy:
        """The default masking policy."""
        return self._default_masking_policy

    @property
    def column_overrides(self) -> Dict[str, MaskingPolicy]:
        """Unmodifiable copy of the per-column override map."""
        return dict(self._column_overrides)

    # ------------------------------------------------------------------
    # Internal: shared match predicate
    # ------------------------------------------------------------------

    def _is_pii(self, field_name: str) -> bool:
        """Return True iff *field_name* is in the column set or matches a regex.

        Single source of truth for both :meth:`classify` and
        :meth:`masking_for` — mirrors Java ``PiiMaskingGovernancePolicy.isPii``
        (PiiMaskingGovernancePolicy.java:197-202).

        Uses ``re.fullmatch`` to mirror Java ``Matcher.matches()``
        (anchored full-string match).
        """
        if field_name in self._pii_columns:
            return True
        for pattern in self._compiled_patterns:
            if pattern.fullmatch(field_name):
                return True
        return False


# ---------------------------------------------------------------------------
# Runtime_checkable assertion (documented, not enforced here)
# ---------------------------------------------------------------------------
# PiiMaskingGovernancePolicy satisfies GovernancePolicy by structural typing
# (duck typing / Protocol). No explicit inheritance is required.
# Verified by: isinstance(PiiMaskingGovernancePolicy(...), GovernancePolicy)
